### PR TITLE
LCOM5 metric implementation

### PR DESCRIPTION
**Pull request by silverCase**

### CODE REVIEW

**Comment by silverCase** [comment]
@acheshkov please, check the test/integration/all.py, if it's right and should be like this
```
EXCLUDE_METRICS = ['M5']
```

**Comment by 0crat** [comment]
This pull request [#288](https://github.com/yegor256/aibolit/issues/288) is assigned to @lyriccoder[/z](https://www.0crat.com/u/lyriccoder), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/4e390f33-c4c5-4dee-898c-80da7013165c); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: test/metrics/lcom4/test_lcom4.py [hunk 1]
```diff
@@ -32,41 +31,56 @@ class TestLCOM4(TestCase):
...
-        lcom4_val = self.pattern.value(Path(self.dir_path, 'Constructor.java'))
+        name = 'Constructor.java'
+        lcom4_val = self.pattern.value(Path(self.dir_path, name))
+        print(name)
         self.assertEqual(lcom4_val, 1)
 
+    def test_scope_with_anonymous(self):
+        name = 'ScopeAnonymous.java'
+        lcom4_val = self.pattern.value(Path(self.dir_path, name))
+        print(name)
```
Comment (lyriccoder) [inline test/metrics/lcom4/test_lcom4.py:48]: why do u have print everywhere? You should remove it

File: test/metrics/lcom5/test_lcom5.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def test_class_with_chain(self):
+        name = 'MethodChain.java'
+        lcom4_val = self.pattern.value(Path(self.dir_path, name))
+        print(name)
+        self.assertEqual(lcom4_val, 7 / 10)
+
+    def test_overloaded_diff(self):
+        name = 'OverloadedDiffComp.java'
+        lcom4_val = self.pattern.value(Path(self.dir_path, name))
+        print(name)
```
Comment (lyriccoder) [inline test/metrics/lcom5/test_lcom5.py:85]: remove print

**Comment by 0crat** [comment]
Job was finished in 2 hours, bonus for fast delivery is possible (see [§36](http://www.zerocracy.com/policy.html#36))

**Comment by 0crat** [comment]
Job audit: `ARC` didn't merge or close PR, performer @lyriccoder[/z](https://www.0crat.com/u/lyriccoder)[/z](https://www.0crat.com/u/lyriccoder) didn't make CR comments

**Comment by 0crat** [comment]
Pull request [#288](https://github.com/yegor256/aibolit/issues/288) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Comment by 0crat** [comment]
This pull request [#288](https://github.com/yegor256/aibolit/issues/288) is assigned to @lyriccoder[/z](https://www.0crat.com/u/lyriccoder), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/67def6c8-1d1b-49c7-a263-00ed693898b5); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: test/metrics/lcom4/test_lcom4.py (hunk 1 above)
Comment (silverCase) [reply test/metrics/lcom4/test_lcom4.py]: Done1

File: aibolit/utils/filter.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+from typing import List, Tuple, Any, Union, TypeVar, Type
+from javalang.tree import ClassDeclaration, InterfaceDeclaration, MethodDeclaration, \
+    MemberReference, FieldDeclaration, MethodInvocation, This, Node, LocalVariableDeclaration
+
+FldExh = Tuple[str, Tuple[str, str]]
```
Comment (acheshkov) [inline aibolit/utils/filter.py:5]: Data types you use in the program have to be declared just one time in a single place.

File: aibolit/utils/filter.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+HasMember = Union[MemberReference, MethodInvocation]
+HasSelector = Union[MemberReference, MethodInvocation, This]
+T = TypeVar('T', bound=Node)
+S = TypeVar('S', HasSelector, HasMember)
+MthNodes = Tuple[tuple, MethodDeclaration]
+Nodes = Tuple[tuple, T]
+AnyField = Union[FieldDeclaration, LocalVariableDeclaration]
+
+
+class Filters:
```
Comment (acheshkov) [inline aibolit/utils/filter.py:16]: I recommend creating unit tests for `Filter` class

File: aibolit/metrics/lcom5/lcom5.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+from aibolit.utils.ast import AST
+from aibolit.utils.cohesiongraph import CohesionGraph
+from aibolit.utils.filter import Filters
+from javalang.tree import Node
+from typing import List
+
+
+class LCOM5:
+
+    coh = CohesionGraph()
```
Comment (acheshkov) [inline aibolit/metrics/lcom5/lcom5.py:32]: why do you declare it as a static attribute?

**Comment by 0crat** [comment]
@lyriccoder[/z](https://www.0crat.com/u/lyriccoder) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by lyriccoder** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/aibolit/issues/288#issuecomment-633608452))

@lyriccoder The user @lyriccoder[/z](https://www.0crat.com/u/lyriccoder) resigned from [#288](https://github.com/yegor256/aibolit/issues/288), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/aibolit/issues/288#issuecomment-633608452))

@lyriccoder There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
PID: 4@66d967ea-fa4a-451b-bad8-e94a595ef078, thread: PQ-CT5UB7XAP
com.jcabi.xml.StrictXML[124] java.lang.IllegalArgumentException: 1 error(s) in XML document: -1:-1: Duplicate unique value [gh:yegor256/aibolit#288] declared for identity constraint of element "speed".
```

[0.53.14](https://github.com/zerocracy/farm): CID: [cab82170-b51b-4538-806d-b4d7c019ed35](https://www.0crat.com/footprint/CT5UB7XAP/cab82170-b51b-4538-806d-b4d7c019ed35), Type: "Order was canceled"

**Comment by 0crat** [comment]
This pull request [#288](https://github.com/yegor256/aibolit/issues/288) is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/dc69943f-2681-4e00-93b6-1f7ce4a58e49); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by paulodamaso** [comment]
@lyriccoder ping

**Comment by lyriccoder** [comment]
I am not a review of it. It has lots conflicts, @acheshkov could you please review it? Seems you were working with this code? It can be familiar code...

**Comment by paulodamaso** [comment]
@lyriccoder Sorry, I have mistaken you as the PR submitter

**Comment by paulodamaso** [comment]
@silverCase Could you please take a look at the comments on https://github.com/yegor256/aibolit/pull/288#pullrequestreview-419744161 ?

**Comment by lyriccoder** [comment]
seems @silverCase does not support this project any longer. We need to find a new person for this code, I suggest @acheshkov since he was reviewing this code for a long time

**Comment by paulodamaso** [comment]
@acheshkov ping

**Comment by acheshkov** [comment]
@paulodamaso, @lyriccoder I have a thought to close this PR

**Comment by acheshkov** [comment]
converted to draft

**Comment by paulodamaso** [comment]
@acheshkov Is the work on this PR still needed? If not, please unassign me from it

**Comment by acheshkov** [comment]
@0crat out

**Comment by acheshkov** [comment]
@paulodamaso nobody is assigned to this task.

**Comment by 0crat** [comment]
> @0crat out ([here](https://github.com/yegor256/aibolit/issues/288#issuecomment-643388788))

@acheshkov GitHub issue is still open, won't close

**Comment by yegor256** [comment]
@0crat status

**Comment by acheshkov** [comment]
@0crat refuse

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/yegor256/aibolit/issues/288#issuecomment-643394801))

@yegor256 This is what I know about this job in CT5UB7XAP, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#288](https://github.com/yegor256/aibolit/issues/288) is [in scope](http://www.zerocracy.com/policy.html#14) for 1mon
  * The role is `REV`
  * The job is assigned to @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) for [17days](http://www.zerocracy.com/policy.html#8)
  * There is a monetary reward attached
  * @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) will get 25% of their hourly rate for this job
  * The job doesn't have any [impediments](http://www.zerocracy.com/policy.html#9)
  * The [budget](http://www.zerocracy.com/policy.html#4) is 15 minutes/points
  * These users are banned and won't be assigned:
    * @silvercase[/z](https://www.0crat.com/u/silvercase): This user reported the ticket
    * @lyriccoder[/z](https://www.0crat.com/u/lyriccoder): User was resigned from the ticket
  * Job [footprint](https://www.0crat.com/footprint/CT5UB7XAP?q=%7Bjob%3A%27gh%3Ayegor256%2Faibolit%23288%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by 0crat** [comment]
> @0crat refuse ([here](https://github.com/yegor256/aibolit/issues/288#issuecomment-643395411))

@acheshkov The user @paulodamaso[/z](https://www.0crat.com/u/paulodamaso) resigned from [#288](https://github.com/yegor256/aibolit/issues/288), please stop working. Reason for job resignation: Order was cancelled

**Comment by 0crat** [comment]
This pull request [#288](https://github.com/yegor256/aibolit/issues/288) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/CT5UB7XAP/8eed7c6b-b059-49cc-8431-bd2c51741456); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @acheshkov[/z](https://www.0crat.com/u/acheshkov) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

**Comment by 0crat** [comment]
Job audit: performer @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)[/z](https://www.0crat.com/u/carlosmiranda) didn't make CR comments

**Comment by 0crat** [comment]
Pull request [#288](https://github.com/yegor256/aibolit/issues/288) was not merged, no payment for `ARC`, see [§28](http://www.zerocracy.com/policy.html#28)

**Pull request closed without merging**